### DISCUSSION TITLE
How do I properly decouple a slice from the main store into its own module?

### DISCUSSION THREAD

**Discussion by michaelschufi** [pmndrs/zustand#1974] [Q&A] [answered]
Hi

First of all, thanks for this great library! I really appreciate it.

However, I'm having trouble decoupling a slice of a store with the main store. How can I do this properly without breaking any rerendering logic?

Here's my attempt so far. The slice is looking good, but I'm not sure how to get the `get` and `set` properly scoped.
See also https://codesandbox.io/p/sandbox/sad-buck-y5vn4t?file=%2Fsrc%2Findex.ts%3A20%2C10
```ts
// bears-slice.ts
import { StateCreator } from "zustand";

export type BearsSlice = {
  counter: number;
  increment: () => void;
};

export type BearsSliceInitialProps = Pick<BearsSlice, "counter">;

export const createBearsSlice =
  (
    initialProps: BearsSliceInitialProps
  ): StateCreator<BearsSlice, [["zustand/immer", never]]> =>
  (set, get) => ({
    ...initialProps,
    increment: () =>
      set((state) => {
        state.counter += 1;
      }),
  });
```

But this breaks atm ...
```ts
// app-store.ts
import { createContext } from "react";
import { createStore } from "zustand";
import { immer } from "zustand/middleware/immer";
import { BearsSlice, BearsSliceInitialProps, createBearsSlice } from "./bears";

export type AppStoreInitialProps = {
  bears: BearsSliceInitialProps;
};

export type AppState = {
  bears: BearsSlice;
};

export type AppStore = ReturnType<typeof createAppStore>;

export const createAppStore = (initProps: AppStoreInitialProps) => {
  return createStore<AppState>()(
    immer<AppState>((set, get, api) => ({
      bears: createBearsSlice(initProps.bears)(
        (bearsState) => set(appState => {
          appState.bears = bearsState
        }),
        () => get().bears,
        api
      ),
    }))
  );
};

export const AppStoreContext = createContext<AppStore | null>(null);
```

I'm looking to decouple a slice like RTK's `createSlice` does. https://redux-toolkit.js.org/api/createSlice

Edit: Also, I would love to have `immer` support as well.

**Comment by dai-shi** [comment, marked as answer]
We had a discussion in the past #178, and the conclusion is that it's out of the scope of this library.
It's an open field for third-party libraries.
One of promising ones is [zustand-lens](https://github.com/dhmk083/dhmk-zustand-lens).

**Reply by michaelschufi** [reply]
Thank you. I'm using that now. Initially, thought that it won't support "Initialize state with props".
Because I'm reluctant of having another dependency that may go out of date with an update, I've also tried it myself in the meantime and I have a simple TypeScript-only solution that also seems to work. I'll post it here later.
